httpserver/dns_text_records.cpp

```diff
@@ -0,0 +1,194 @@
+#include "../external/json.hpp"
+#include "dns_text_records.h"
+#include "version.h"
+#include "pow.hpp"
+#include <resolv.h>
+
+#include <boost/algorithm/string.hpp>
+
+using json = nlohmann::json;
+
+static constexpr char POW_DIFFICULTY_URL[] = "sentinel.messenger.loki.network";
+static constexpr char LATEST_VERSION_URL[] = "storage.version.loki.network";
+
+namespace loki {
+
+namespace dns {
+
+static std::string get_dns_record(const char* url, std::error_code& ec) {
+
+    std::string data;
+    unsigned char query_buffer[1024] = {};
+
+    // don't want to assume that ec has default value
+    ec = std::error_code{};
+
+    int response =
+        res_query(url, ns_c_in, ns_t_txt, query_buffer, sizeof(query_buffer));
+    
+    if (response == -1) {
+        LOKI_LOG(warn, "res_query failed while retrieving PoW");
+        ec = std::make_error_code(std::errc::bad_message);
+        return data;
+    }
+
+    ns_msg nsMsg;
+
+    if (ns_initparse(query_buffer, response, &nsMsg) == -1) {
+        LOKI_LOG(warn, "ns_initparse failed while retrieving PoW");
+        ec = std::make_error_code(std::errc::bad_message);
+        return data;
+    }
+
+    // We get back a sequence of N...[N...] values where N is a byte indicating
+    // the length of the immediately following ... data.
+    auto count = ns_msg_count(nsMsg, ns_s_an);
+
+    data.reserve(255 * count);
+    for (int i = 0; i < count; i++) {
+        ns_rr rr;
+        if (ns_parserr(&nsMsg, ns_s_an, i, &rr) == -1) {
+            LOKI_LOG(warn, "ns_parserr failed while parsing PoW data");
+            ec = std::make_error_code(std::errc::bad_message);
+            return data;
+        }
+        auto* rdata = ns_rr_rdata(rr);
+        data.append(reinterpret_cast<const char*>(rdata + 1), rdata[0]);
+    }
+
+    return data;
+}
+
+std::vector<pow_difficulty_t> query_pow_difficulty(std::error_code& ec) {
+    LOKI_LOG(debug, "Querying PoW difficulty...");
+
+    std::vector<pow_difficulty_t> new_history;
+    const std::string data = get_dns_record(POW_DIFFICULTY_URL, ec);
+    if (ec) {
+        return new_history;
+    }
+
+    try {
+        const json history = json::parse(data, nullptr, true);
+        for (const auto& el : history.items()) {
+            const std::chrono::milliseconds timestamp(std::stoul(el.key()));
+            const int difficulty = el.value().get<int>();
+            new_history.push_back(pow_difficulty_t{timestamp, difficulty});
+        }
+        return new_history;
+    } catch (const std::exception& e) {
+        LOKI_LOG(warn, "JSON parsing of PoW data failed: {}", e.what());
+        ec = std::make_error_code(std::errc::bad_message);
+        return new_history;
+    }
+}
+
+static std::string query_latest_version() {
+
```

Comment:
Do we want to throw a log here as well like above?

Reply:
done!